### PR TITLE
init FFI

### PR DESCRIPTION
**Pull request by BusyJay**
This pr initializes the project and adds needed FFIs.

To try it out locally, you will need to install GRPC native libraries. Only 1.2.5 is tested.

@siddontang @overvenus @disksing PTAL

### CODE REVIEW

File: grpc-sys/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    pub fn grpc_server_destroy(server: *mut GrpcServer);
+}
+
+// TODO: more tests.
+#[cfg(test)]
+mod tests {
+    use std::ptr;
+
+    #[test]
+    fn smoke() {
```
Comment (disksing) [inline grpc-sys/src/lib.rs]: What does smoke mean?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+target
+Cargo.lock
```
Comment (overvenus) [inline .gitignore]: Please add `*.rs.bk`

File: Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+version = "0.1.0"
+
+[dependencies]
+grpc-sys = { path = "grpc-sys", version = "0.1" }
+libc = "0.2"
+futures = "0.1"
+protobuf = "1.2"
+clippy = { version = "*", optional = true }
+
+[workspace]
```
Comment (overvenus) [inline Cargo.toml]: Why not add `grpc-sys` ?

File: grpc-sys/build.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+fn main() {
+    // TODO: support static link
+    gcc::Config::new()
+        .file("grpc_wrap.c")
+        .flag("-fPIC")
+        .flag("-O2")
+        .compile("libgrpc_wrap.a");
+
+    println!("cargo:rustc-link-lib=grpc");
```
Comment (overvenus) [inline grpc-sys/build.rs]: Do we have to install gRPC manually?

File: grpc-sys/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+#[repr(C)]
+pub struct GprTimespec {
+    tv_sec: int64_t,
+    tv_nsec: int32_t,
+    clock_type: GprClockType,
+}
+
+impl From<Option<Duration>> for GprTimespec {
```
Comment (overvenus) [inline grpc-sys/src/lib.rs]: Why not `From<Duration>`? As for `Option<Duration>`, we can `option.map(From::from).unwrap()`.

File: .gitignore (hunk 2 above)
Comment (overvenus) [inline .gitignore]: And `*.rs.fmt`

File: Cargo.toml (hunk 3 above)
Comment (BusyJay) [inline Cargo.toml]: It doesn't need to be added explicitly.

File: grpc-sys/build.rs (hunk 4 above)
Comment (BusyJay) [inline grpc-sys/build.rs]: Yes, as the description of the pr says.

File: grpc-sys/src/lib.rs (hunk 5 above)
Comment (BusyJay) [inline grpc-sys/src/lib.rs]: Because I want to convert `None` to `Infinite`.

File: .gitignore (hunk 2 above)
Comment (BusyJay) [inline .gitignore]: What's `*.rs.fmt`?

File: Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "grpc"
+version = "0.1.0"
+
+[dependencies]
+grpc-sys = { path = "grpc-sys", version = "0.1" }
+libc = "0.2"
```
Comment (andelf) [inline Cargo.toml]: I often use ``std::os::raw``. 🤣

File: grpc-sys/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    Unauthenticated = 16,
+    ResourceExhausted = 8,
+    FailedPrecondition = 9,
+    Aborted = 10,
+    OutOfRange = 11,
+    Unimplemented = 12,
+    Internal = 13,
+    Unavailable = 14,
+    DataLoss = 15,
+    DoNotUse = -1,
```
Comment (andelf) [inline grpc-sys/src/lib.rs]: unnecessary?

File: .gitignore (hunk 2 above)
Comment (overvenus) [inline .gitignore]: My VSCode sometimes outputs a `*.rs.fmt` file.

File: grpc-sys/src/lib.rs (hunk 5 above)
Comment (overvenus) [inline grpc-sys/src/lib.rs]: How about `map_or_else` ?

File: grpc-sys/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    Shutdown,
+}
+
+#[repr(C)]
+pub enum GrpcCompressionLevel {
+    None = 0,
+    Low,
+    Med,
+    High,
+    Count,
```
Comment (andelf) [inline grpc-sys/src/lib.rs]: unnecessary. Count denotes last unused entry.

**Comment by siddontang** [comment]
do we have any tool to format the c code?

**Comment by disksing** [comment]
LGTM.

File: grpc-sys/grpc_wrap.c [hunk 9]
```diff
@@ -0,0 +1,864 @@
...
+  gpr_free(ctx);
+}
+
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_batch_context_set_tag(grpcwrap_batch_context *ctx, void *tag) {
+  ctx->tag = tag;
+}
+
+GPR_EXPORT void *GPR_CALLTYPE
+grpcwrap_batch_context_take_tag(grpcwrap_batch_context *ctx) {
```
Comment (siddontang) [inline grpc-sys/grpc_wrap.c]: no get tag API?

**Comment by disksing** [comment]
When I type 'cargo test', it only runs tests inside 'src', could you make it run tests inside 'grpc-sys' too?

File: grpc-sys/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+
+#[repr(C)]
+pub struct GprTimespec {
+    tv_sec: int64_t,
+    tv_nsec: int32_t,
+    clock_type: GprClockType,
+}
+
+impl GprTimespec {
+    pub fn inf() -> GprTimespec {
```
Comment (andelf) [inline grpc-sys/src/lib.rs]: bad naming.
there's a gpr_inf_past()

File: rustfmt.toml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+reorder_imports = true
+reorder_imported_names = true
```
Comment (andelf) [inline rustfmt.toml]: EOF

File: grpc-sys/src/lib.rs (hunk 10 above)
Comment (siddontang) [inline grpc-sys/src/lib.rs]: inf_future +1

**Comment by siddontang** [comment]
Rest LGTM

File: grpc-sys/grpc_wrap.c (hunk 9 above)
Comment (BusyJay) [inline grpc-sys/grpc_wrap.c]: No, the object will retake its ownership once cast back from a pointer.

**Comment by BusyJay** [comment]
You can run all tests via `cargo test --all`.

**Comment by BusyJay** [comment]
I formatted the c code with `clang-format -style=Google -i grpc_wrap.c`.

**Comment by siddontang** [comment]
you can support this with `make format`

**Comment by BusyJay** [comment]
Maybe next pr.

**Comment by siddontang** [comment]
PTAL @overvenus

**Pull request merged**